File: September2022/Java/MaximumLengthOfRepeatedSubarray.java
// @saorav21994

class Solution {
    public int findLength(int[] nums1, int[] nums2) {
        int l1 = nums1.length;
        int l2 = nums2.length;
        int [][] dp = new int[l1][l2];
        
        // using top-down approach dynamic-programming
        // if equal, (diagonal prev + 1)
        
        int res = 0;
        
        // first row initialize
        for (int i = 0; i < l2; i++) {
            if (nums1[0] == nums2[i]) {
                dp[0][i] = 1;
                res = Math.max(res, dp[0][i]);
            }
        }
        // first column initialize, dp[0][0] already accounted for
        for (int i = 1; i < l1; i++) {
            if (nums2[0] == nums1[i]) {
                dp[i][0] = 1;
                res = Math.max(res, dp[i][0]);
            }
        }
        // for (int []r : dp) {
        //     for (int c : r) {
        //         System.out.print(c +" ");
        //     }
        //     System.out.println();
        // }
        // System.out.println();
        
        // fill out rest elements
        for (int i = 1; i < l1; i++) {
            for (int j = 1; j < l2; j++) {
                if (nums1[i] == nums2[j]) {
                    dp[i][j] = dp[i-1][j-1]+1;
                    res = Math.max(dp[i][j], res);
                }

            }
        }
        // for (int []r : dp) {
        //     for (int c : r) {
        //         System.out.print(c +" ");
        //     }
        //     System.out.println();
        // }
        return res;
    }
}


============================ Clean code ============================
    // author: anuj0503
    class Solution {
    public int findLength(int[] nums1, int[] nums2) {
        int n = nums1.length;
        int m = nums2.length;
        int[][] dp = new int[n+1][m+1];
        int ans = 0;
        
        for(int i = 1; i <= n; i++)
            for(int j = 1; j <= m; j++){
                if(nums1[i-1] == nums2[j-1]){
                    dp[i][j] = dp[i-1][j-1] + 1;
                    ans = Math.max(ans, dp[i][j]);
                }
            }
        
        return ans;
    }
}

File: September2022/C++/MaximumLengthOfRepeatedSubarray.cpp
// https://leetcode.com/problems/maximum-length-of-repeated-subarray/

// Dynamic Programming - Tabulation ✅
class Solution {
public:
    // TC: O(n * m)
    // SC: O(n * m) -> dp array
    int findLength(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();
        
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
        int ans = 0;
        
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (nums1[i - 1] == nums2[j - 1]) {
                    dp[i][j] = 1 + dp[i - 1][j - 1];
                    ans = max(ans, dp[i][j]);
                } else {
                    dp[i][j] = 0;
                }
            }
        }
        
        return ans;
    }
};


// Dynamic Programming - Tabulation + Space Optimization ✅
class Solution {
public:
    // TC: O(n * m)
    // SC: O(2 * m)
    int findLength(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();
        
        vector<int> prev(m + 1, 0);
        vector<int> curr(m + 1, 0);
        int ans = 0;
        
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (nums1[i - 1] == nums2[j - 1]) {
                    curr[j] = 1 + prev[j - 1];
                    ans = max(ans, curr[j]);
                } else {
                    curr[j] = 0;
                }
            }
            prev = curr;
        }
        
        return ans;
    }
};
